src/types/job.rs

```diff
@@ -41,6 +41,22 @@ pub struct JobDescriptor {
     pub num_incomplete: u32,
 }
 
+/// Either a list of package descriptors or a list of PURLs.
+#[derive(
+    Clone, Debug, Deserialize, Eq, Hash, JsonSchema, Ord, PartialEq, PartialOrd, Serialize,
+)]
+#[serde(untagged)]
+pub enum PackageDescriptorsOrPurls {
```

Comment:
Not really a fan of the enum name here. Maybe `PackageIdentifier` would be more descriptive of its purpose rather than its contents?

Reply:
I think `PackageIdentifier` would be confusing. We currently have `PackageDescriptor` `PackageVersionLocator` and `PackageSpecifier` all referring to different ways of combining type+name+version. `PackageDescriptorsOrPurls` is much more descriptive, and avoids the problem of coming up with more names in API where there are interfaces that deal with two other `XOrPurl` types (luckily only one of those needs to be exposed).

Maybe we could use [either](https://crates.io/crates/either) or define our own either enum.

Reply:
Also `PackageIdentifier` would need to be `PackageIdentifiers` unless we switch to accepting mixed lists where each element is either a purl or a package descriptor, which seems unnecessary since we prefer one form over the other and only support both for compatibility. I think it's cleaner to handle the case where they're all the same form, but it probably doesn't make too much of a difference.

Reply:
The problem with `PackageDescriptorsOrPurls` is that it doesn't describe the purpose of the enum but what variants it has. This will fall apart as soon as another variant is added. That's just generally a bad way to name things for the sake of maintenance.

Reply:
There aren't supposed to be new variants added later. This is a step towards replacing one variant with another variant but we need to support both for a while because we can't switch everything at once unless we make completely new API endpoints.

I don't see a problem with naming the enum after its variants because if we didn't have the enum, we would be directly using its variants. It'd be different if it were named `PackageDescriptorsOrStrings`.

If we keep inventing new generic types for specifying packages we're going to run out of words. We could make it two enums, `JobDescriptorPackages` and `SubmitPackageDependencies`.

Reply:
> if we didn't have the enum, we would be directly using its variants

I don't understand this point...

Naming the enum after the variants means that we will be writing values like this: `PackageDescriptorOrPurls::PackageDescriptors(packages)`

> If we keep inventing new generic types for specifying packages we're going to run out of words.

We don't have to though, right? This type is going to take over the job of `PackageDescriptor` in the API. We could just name this type `PackageDescriptor` and use different modules for the stages of this transition. Here is an incredibly abbreviated example:

```rust
pub mod legacy {
    pub struct PackageDescriptor { /* current implementation */ }
}

pub mod purl {
    pub type PackageDescriptor = String;
}

pub mod maybe_purl {
    pub enum PackageDescriptor {
        Legacy(super::legacy::PackageDescriptor),
        Purl(String),
    }
}
```

Reply:
`legacy::PackageDescriptor` and `purl::PackageDescriptor` and `maybe_purl::PackageDescriptor` seems confusing. In the current code, a `PackageDescriptor` is a type with three fields. This changes `PackageDescriptor` into just a concept that refers to a package. It's still necessary to talk about concrete `PackageDescriptor`s, so I'd much rather call it `PackageDescriptorObject` or something if that's the way we're going. I don't want code for converting between `PackageDescriptor`s and `PackageDescriptor`s or compile errors because `PackageDescriptor` was expected and `PackageDescriptor` was found.

`PackageDescriptor` the enum does not replace `PackageDescriptor` the struct. In the API interface and in persisted data both are accepted for version compatibility reasons, but in most places it's specifically one or the other. I just encountered a problem where by changing `PackageDescriptor` to be an enum, the lockfile parser crate starts outputting this enum. Allowing it to be either variant forces the surrounding code to expect PURLs now, even though they won't be encountering any yet, and then once the lockfile parser crate starts outputting PURLs the code will still be expecting to sometimes get the structs instead.

There's also a problem with using `PackageDescriptor::Purl` for anything besides persistence and RPC. `String` is great for the interface because it's simple, adds no dependencies, and doesn't get in the way. However, inside the API we want parsed PURL objects that have fields. The plan is to first accept PURLs and convert them to `PackageVersionLocator` (very similar to `PackageDescriptor` the struct) so the interface is there and works the same as before except now with PURL support, then to go through and make the API natively use PURL, not `PackageDescriptor` enums or `PackageDescriptor` structs or `PackageVersionLocator`s.

Reply:
I didn't even notice the usability problem of trying to write `PackageDescriptor::Legacy(PackageDescriptor { ... })`. I guess it works if everything has `From` impls and you only ever need either the enum or the struct in a single module.

Reply:
I still think `PackageDescriptorsOrPurls` or `PackageDescriptorOrPurl` is the best option out of these. `PackageDescriptor` causes too many problems. `PackageIdentifier` works, but to me the idea of naming the type after what it's for rather than what it is only creates confusion, especially with the way we use `PackageDescriptor` to describe a package version, an unversioned package name, or a dependency specification, and we also use `PackageSpecifier` interchangeably for the same purposes. The distinction from `PackageVersionLocator` is a little clearer because `PackageVersionLocator` cannot be used to describe an unversioned package name, but it is otherwise the same thing with a different package type enum and name normalization rules.

From trying out changing `PackageDescriptor` to an enum, I think there might be another option: do nothing. Making this interface change affects only `cli::api::PhylumApi::submit_request`. `SubmitPackageRequest` is apparently the only struct defined in `phylum_types` that contains `PackageDescriptor`s and is used by CLI to send requests to the API. API can just define its own version of this struct that accepts either `PackageSpecifier`s or PURLs.

This has multiple benefits:
- We don't need to introduce another common type for referencing packages.
- We don't need to go through all the consumers of `phylum_types` updating them to use a different `PackageDescriptor` or to use an enum that might contain a PURL now, then again to natively use PURLs but still wrapped with an enum where they might be `PackageDescriptor`s, then again to only use PURLs.
- It breaks this circular dependency where adding a new feature or removing a deprecated feature requires introducing optional values that are not supposed to be used, deprecated optional fields that must be set, enum variants that must not be sent but must be accepted on receipt, etc. Only API will be concerned with accepting either type in requests and then later only CLI will be concerned with accepting either type in responses if the same endpoint is going to start returning the new format (pushing this decision down the road). (Although it doesn't seem like CLI ever actually uses the returned `PackageDescriptor`s today and maybe we could just remove them. In the places CLI/UI does care, the PURL is being added into a larger struct describing the package.)

The two other CLI calls that send package references to the API either define their own struct or just send a `PackageSpecifier` directly. They would not need to change.

I'll try this.